File: frontend-api/src/main/java/no/nav/melosys/tjenester/gui/dto/BehandlingsresultatDto.java
package no.nav.melosys.tjenester.gui.dto;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import no.nav.melosys.domain.Behandlingsresultat;
import no.nav.melosys.domain.BehandlingsresultatBegrunnelse;
import no.nav.melosys.domain.Kontrollresultat;
import no.nav.melosys.domain.kodeverk.begrunnelser.Kontroll_begrunnelser;
import no.nav.melosys.domain.kodeverk.behandlinger.Behandlingsresultattyper;

public class BehandlingsresultatDto {
    private final String behandlingsresultatTypeKode;
    private final List<String> begrunnelseKoder;
    private final String begrunnelseFritekst;
    private final String innledningFritekst;
    private final String utfallRegistreringUnntak;
    private final String utfallUtpeking;
    private final String vedtakstype;
    private final List<String> kontrollresultatBegrunnelseKoder;

    private BehandlingsresultatDto(Behandlingsresultattyper behandlingsresultatTypeKode,
                                   String begrunnelseFritekst,
                                   String innledningFritekst,
                                   String utfallRegistreringUnntak,
                                   String utfallUtpeking,
                                   String vedtakstype,
                                   List<String> kontrollresultatBegrunnelseKoder) {
        this.behandlingsresultatTypeKode = behandlingsresultatTypeKode.getKode();
        this.innledningFritekst = innledningFritekst;
        this.utfallUtpeking = utfallUtpeking;
        this.begrunnelseKoder = new ArrayList<>();
        this.begrunnelseFritekst = begrunnelseFritekst;
        this.utfallRegistreringUnntak = utfallRegistreringUnntak;
        this.vedtakstype = vedtakstype;
        this.kontrollresultatBegrunnelseKoder = kontrollresultatBegrunnelseKoder;
    }

    public static BehandlingsresultatDto av(Behandlingsresultat resultat) {
        BehandlingsresultatDto dto = new BehandlingsresultatDto(
            resultat.getType(),
            resultat.getBegrunnelseFritekst(),
            resultat.getInnledningFritekst(),
            resultat.getUtfallRegistreringUnntak() != null ? resultat.getUtfallRegistreringUnntak().getKode() : null,
            resultat.getUtfallUtpeking() != null ? resultat.getUtfallUtpeking().getKode() : null,
            resultat.getVedtakMetadata() != null ? resultat.getVedtakMetadata().getVedtakstype().getKode() : null,
            resultat.getKontrollresultater().stream().map(Kontrollresultat::getBegrunnelse).map(Kontroll_begrunnelser::getKode).collect(Collectors.toList())
        );

        resultat.getBehandlingsresultatBegrunnelser().stream()
            .map(BehandlingsresultatBegrunnelse::getKode)
            .forEach(dto.getBegrunnelseKoder()::add);

        if (resultat.getVedtakMetadata() != null && resultat.getVedtakMetadata().getRevurderBegrunnelse() != null) {
            dto.getBegrunnelseKoder().add(resultat.getVedtakMetadata().getRevurderBegrunnelse());
        }
        return dto;
    }

    public String getBehandlingsresultatTypeKode() {
        return behandlingsresultatTypeKode;
    }

    public List<String> getBegrunnelseKoder() {
        return begrunnelseKoder;
    }

    public String getBegrunnelseFritekst() {
        return begrunnelseFritekst;
    }

    public String getInnledningFritekst() {
        return innledningFritekst;
    }

    public String getUtfallRegistreringUnntak() {
        return utfallRegistreringUnntak;
    }

    public String getUtfallUtpeking() {
        return utfallUtpeking;
    }

    public String getVedtakstype() {
        return vedtakstype;
    }

    public List<String> getKontrollresultatBegrunnelseKoder() {
        return kontrollresultatBegrunnelseKoder;
    }
}

File: frontend-api/src/main/kotlin/no/nav/melosys/tjenester/gui/dto/BehandlingsresultatDto.kt
package no.nav.melosys.tjenester.gui.dto

import no.nav.melosys.domain.Behandlingsresultat
import no.nav.melosys.domain.BehandlingsresultatBegrunnelse
import no.nav.melosys.domain.Kontrollresultat
import no.nav.melosys.domain.kodeverk.begrunnelser.Kontroll_begrunnelser
import no.nav.melosys.domain.kodeverk.behandlinger.Behandlingsresultattyper
import java.util.stream.Collectors

class BehandlingsresultatDto private constructor(
    behandlingsresultatTypeKode: Behandlingsresultattyper,
    val begrunnelseFritekst: String?,
    val innledningFritekst: String?,
    val trygdeavgiftFritekst: String?,
    val nyVurderingBakgrunn: String?,
    val utfallRegistreringUnntak: String?,
    val utfallUtpeking: String?,
    val vedtakstype: String?,
    val kontrollresultatBegrunnelseKoder: List<String>?,
    val fakturaserieReferanse: String?
) {
    val behandlingsresultatTypeKode: String = behandlingsresultatTypeKode.kode
    @JvmField
    val begrunnelseKoder: MutableList<String> = ArrayList()

    companion object {
        @JvmStatic
        fun av(resultat: Behandlingsresultat): BehandlingsresultatDto {
            val dto = BehandlingsresultatDto(
                resultat.type,
                resultat.begrunnelseFritekst,
                resultat.innledningFritekst,
                resultat.trygdeavgiftFritekst,
                resultat.nyVurderingBakgrunn,
                if (resultat.utfallRegistreringUnntak != null) resultat.utfallRegistreringUnntak.kode else null,
                if (resultat.utfallUtpeking != null) resultat.utfallUtpeking.kode else null,
                if (resultat.vedtakMetadata != null && resultat.vedtakMetadata.vedtakstype != null
                ) resultat.vedtakMetadata.vedtakstype.kode else null,
                resultat.kontrollresultater.stream().map { obj: Kontrollresultat -> obj.begrunnelse }
                    .map { obj: Kontroll_begrunnelser -> obj.kode }.collect(Collectors.toList()),
                resultat.fakturaserieReferanse
            )

            resultat.behandlingsresultatBegrunnelser.stream()
                .map { obj: BehandlingsresultatBegrunnelse -> obj.kode }
                .forEach { e: String -> dto.begrunnelseKoder.add(e) }

            if (resultat.nyVurderingBakgrunn != null) {
                dto.begrunnelseKoder.add(resultat.nyVurderingBakgrunn)
            }
            return dto
        }
    }
}
